protocols/autonatv2/src/server/handler/dial_request.rs

```diff
@@ -0,0 +1,355 @@
+use std::{
+    io,
+    sync::Arc,
+    task::{Context, Poll},
+    time::Duration,
+};
+
+use either::Either;
+use futures::{
+    channel::{mpsc, oneshot},
+    AsyncRead, AsyncWrite, SinkExt, StreamExt,
+};
+use futures_bounded::FuturesSet;
+use libp2p_core::{
+    upgrade::{DeniedUpgrade, ReadyUpgrade},
+    Multiaddr,
+};
+use libp2p_identity::PeerId;
+use libp2p_swarm::{
+    handler::{ConnectionEvent, FullyNegotiatedInbound, ListenUpgradeError},
+    ConnectionHandler, ConnectionHandlerEvent, StreamProtocol, SubstreamProtocol,
+};
+use rand_core::RngCore;
+
+use crate::{
+    generated::structs::{mod_DialResponse::ResponseStatus, DialStatus},
+    request_response::{DialDataRequest, DialRequest, DialResponse, Request, Response},
+    Nonce,
+};
+use crate::{request_response::Coder, REQUEST_PROTOCOL_NAME};
+
+#[derive(Clone, Debug)]
+pub struct StatusUpdate {
+    pub all_addrs: Vec<Multiaddr>,
+    pub tested_addr: Option<Multiaddr>,
```

Comment:
When can this be `None`? Wouldn't it make sense to just not emit an update then?

Reply:
We could do that, but there is a case: It can happen when the client send a `DialDataResponse` instead of the expected `DialRequest`.
If you want, we can just not emit an update then.

Reply:
We can handle these kind of protocol violations by emitting logs.

Reply:
0k we are doing that now